File: app/src/main/java/com/josycom/mayorjay/flowoverstack/ui/SearchActivity.java
package com.josycom.mayorjay.flowoverstack.ui;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.widget.NestedScrollView;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import com.josycom.mayorjay.flowoverstack.R;
import com.josycom.mayorjay.flowoverstack.adapters.SearchAdapter;
import com.josycom.mayorjay.flowoverstack.databinding.ActivitySearchBinding;
import com.josycom.mayorjay.flowoverstack.model.Owner;
import com.josycom.mayorjay.flowoverstack.model.Question;
import com.josycom.mayorjay.flowoverstack.util.DateUtil;
import com.josycom.mayorjay.flowoverstack.util.StringConstants;
import com.josycom.mayorjay.flowoverstack.viewmodel.SearchViewModel;

import java.util.List;
import java.util.Objects;

import static com.josycom.mayorjay.flowoverstack.util.StringConstants.EXTRA_AVATAR_ADDRESS;
import static com.josycom.mayorjay.flowoverstack.util.StringConstants.EXTRA_QUESTION_ANSWERS_COUNT;
import static com.josycom.mayorjay.flowoverstack.util.StringConstants.EXTRA_QUESTION_DATE;
import static com.josycom.mayorjay.flowoverstack.util.StringConstants.EXTRA_QUESTION_FULL_TEXT;
import static com.josycom.mayorjay.flowoverstack.util.StringConstants.EXTRA_QUESTION_ID;
import static com.josycom.mayorjay.flowoverstack.util.StringConstants.EXTRA_QUESTION_NAME;
import static com.josycom.mayorjay.flowoverstack.util.StringConstants.EXTRA_QUESTION_OWNER_LINK;
import static com.josycom.mayorjay.flowoverstack.util.StringConstants.EXTRA_QUESTION_TITLE;

public class SearchActivity extends AppCompatActivity {

    private ActivitySearchBinding mActivitySearchBinding;
    private String mSearchInput;
    private List<Question> mQuestions;
    private SearchViewModel mSearchViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mActivitySearchBinding = ActivitySearchBinding.inflate(getLayoutInflater());
        setContentView(mActivitySearchBinding.getRoot());

        mActivitySearchBinding.rvSearchResults.setLayoutManager(new LinearLayoutManager(this));
        mActivitySearchBinding.rvSearchResults.setHasFixedSize(true);
        mActivitySearchBinding.rvSearchResults.setItemAnimator(new DefaultItemAnimator());

        mActivitySearchBinding.searchScrollUpFab.setVisibility(View.INVISIBLE);
        mActivitySearchBinding.searchNestedScrollview.setOnScrollChangeListener((NestedScrollView.OnScrollChangeListener) (v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
            if (scrollY > 0) {
                mActivitySearchBinding.searchScrollUpFab.setVisibility(View.VISIBLE);
            } else {
                mActivitySearchBinding.searchScrollUpFab.setVisibility(View.INVISIBLE);
            }
        });
        mActivitySearchBinding.searchScrollUpFab.setOnClickListener(v -> mActivitySearchBinding.searchNestedScrollview.scrollTo(0, 0));

        View.OnClickListener mOnClickListener = v -> {
            RecyclerView.ViewHolder viewHolder = (RecyclerView.ViewHolder) v.getTag();
            int position = viewHolder.getAdapterPosition();
            Intent answerActivityIntent = new Intent(getApplicationContext(), AnswerActivity.class);
            Question currentQuestion = mQuestions.get(position);
            Owner questionOwner = currentQuestion.getOwner();

            answerActivityIntent.putExtra(EXTRA_QUESTION_TITLE, currentQuestion.getTitle());
            answerActivityIntent.putExtra(EXTRA_QUESTION_NAME, questionOwner.getDisplayName());
            answerActivityIntent.putExtra(EXTRA_QUESTION_DATE,
                    DateUtil.toNormalDate(currentQuestion.getCreationDate()));
            answerActivityIntent.putExtra(EXTRA_QUESTION_FULL_TEXT, currentQuestion.getBody());
            answerActivityIntent.putExtra(EXTRA_AVATAR_ADDRESS, questionOwner.getProfileImage());
            answerActivityIntent.putExtra(EXTRA_QUESTION_ANSWERS_COUNT, currentQuestion.getAnswerCount());
            answerActivityIntent.putExtra(EXTRA_QUESTION_ID, currentQuestion.getQuestionId());
            answerActivityIntent.putExtra(EXTRA_QUESTION_OWNER_LINK, questionOwner.getLink());

            startActivity(answerActivityIntent);
            overridePendingTransition(R.anim.fade_in_anim, R.anim.fade_out_anim);
        };

        // What happens when the search button is clicked
        mActivitySearchBinding.searchButton.setOnClickListener(v -> {
            if (TextUtils.isEmpty(Objects.requireNonNull(mActivitySearchBinding.searchTextInputEditText.getText()).toString())) {
                mActivitySearchBinding.searchTextInputEditText.setError(getString(R.string.type_a_search_query));
            } else {
                mSearchInput = Objects.requireNonNull(mActivitySearchBinding.searchTextInputEditText.getText()).toString();
                InputMethodManager inputMethodManager = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
                if (inputMethodManager != null) {
                    inputMethodManager.hideSoftInputFromWindow(Objects.requireNonNull(getCurrentFocus()).getWindowToken(), 0);
                }
                makeSearch();
            }
        });
        final SearchAdapter searchAdapter = new SearchAdapter();
        mSearchViewModel = new ViewModelProvider(this).get(SearchViewModel.class);

        mSearchViewModel.getResponseLiveData().observe(this, searchResponse -> {
            switch (searchResponse.networkState) {
                case StringConstants.LOADING:
                    onLoading();
                    break;
                case StringConstants.LOADED:
                    onLoaded();
                    mQuestions = searchResponse.questions;
                    searchAdapter.setQuestions(searchResponse.questions);
                    break;
                case StringConstants.NO_MATCHING_RESULT:
                    onNoMatchingResult();
                    break;
                case StringConstants.FAILED:
                    onError();
                    break;
            }
        });
        mActivitySearchBinding.rvSearchResults.setAdapter(searchAdapter);
        searchAdapter.setOnClickListener(mOnClickListener);
    }

    private void makeSearch() {
        mSearchViewModel.setQuery(mSearchInput);
    }

    private void onLoading() {
        mActivitySearchBinding.searchPbFetchData.setVisibility(View.VISIBLE);
        mActivitySearchBinding.rvSearchResults.setVisibility(View.INVISIBLE);
        mActivitySearchBinding.searchTvError.setVisibility(View.INVISIBLE);
    }

    private void onLoaded() {
        mActivitySearchBinding.searchPbFetchData.setVisibility(View.INVISIBLE);
        mActivitySearchBinding.rvSearchResults.setVisibility(View.VISIBLE);
        mActivitySearchBinding.searchTvError.setVisibility(View.INVISIBLE);
    }

    private void onNoMatchingResult() {
        mActivitySearchBinding.searchPbFetchData.setVisibility(View.INVISIBLE);
        mActivitySearchBinding.rvSearchResults.setVisibility(View.INVISIBLE);
        mActivitySearchBinding.searchTvError.setVisibility(View.VISIBLE);
        mActivitySearchBinding.searchTvError.setText(R.string.no_matching_result);
    }

    private void onError() {
        mActivitySearchBinding.searchPbFetchData.setVisibility(View.INVISIBLE);
        mActivitySearchBinding.rvSearchResults.setVisibility(View.INVISIBLE);
        mActivitySearchBinding.searchTvError.setVisibility(View.VISIBLE);
        mActivitySearchBinding.searchTvError.setText(R.string.search_error_message);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        overridePendingTransition(R.anim.fade_in_anim, R.anim.fade_out_anim);
    }
}

File: app/src/main/java/com/josycom/mayorjay/flowoverstack/ui/activity/SearchActivity.kt
package com.josycom.mayorjay.flowoverstack.ui.activity

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import androidx.appcompat.app.AppCompatActivity
import androidx.core.widget.NestedScrollView
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.josycom.mayorjay.flowoverstack.R
import com.josycom.mayorjay.flowoverstack.databinding.ActivitySearchBinding
import com.josycom.mayorjay.flowoverstack.model.Question
import com.josycom.mayorjay.flowoverstack.network.SearchResponse
import com.josycom.mayorjay.flowoverstack.ui.adapters.SearchAdapter
import com.josycom.mayorjay.flowoverstack.ui.viewmodel.CustomSearchViewModelFactory
import com.josycom.mayorjay.flowoverstack.ui.viewmodel.SearchViewModel
import com.josycom.mayorjay.flowoverstack.util.AppConstants
import com.josycom.mayorjay.flowoverstack.util.AppUtils
import dagger.android.AndroidInjection
import javax.inject.Inject

class SearchActivity : AppCompatActivity() {

    private lateinit var binding: ActivitySearchBinding
    private var searchInput: String? = null
    private var questions: List<Question>? = null
    private lateinit var searchViewModel: SearchViewModel
    @Inject
    lateinit var viewModelFactory: CustomSearchViewModelFactory
    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        binding = ActivitySearchBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModelFactory.setInputs(AppConstants.FIRST_PAGE, AppConstants.SEARCH_PAGE_SIZE)
        binding.apply {
            rvSearchResults.layoutManager = LinearLayoutManager(this@SearchActivity)
            rvSearchResults.setHasFixedSize(true)
            rvSearchResults.itemAnimator = DefaultItemAnimator()
            searchScrollUpFab.visibility = View.INVISIBLE
            searchNestedScrollview.setOnScrollChangeListener { _: NestedScrollView?, _: Int, scrollY: Int, _: Int, _: Int ->
                if (scrollY > 0) {
                    searchScrollUpFab.visibility = View.VISIBLE
                } else {
                    searchScrollUpFab.visibility = View.INVISIBLE
                }
            }
            searchScrollUpFab.setOnClickListener { searchNestedScrollview.scrollTo(0, 0) }
        }

        binding.searchButton.setOnClickListener {
            if (TextUtils.isEmpty(binding.searchTextInputEditText.text.toString())) {
                binding.searchTextInputEditText.error = getString(R.string.type_a_search_query)
            } else {
                searchInput = binding.searchTextInputEditText.text.toString().trim()
                val inputMethodManager = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                inputMethodManager.hideSoftInputFromWindow(currentFocus!!.windowToken, 0)
                setQuery()
                binding.ivLookup.visibility = View.INVISIBLE
            }
        }
        val searchAdapter = SearchAdapter()
        searchViewModel = ViewModelProvider(this, viewModelFactory).get(SearchViewModel::class.java)
        searchViewModel.responseLiveData.observe(this, { searchResponse: SearchResponse ->
            when (searchResponse.networkState) {
                AppConstants.LOADING -> onLoading()
                AppConstants.LOADED -> {
                    onLoaded()
                    questions = searchResponse.questions
                    searchAdapter.setQuestions(searchResponse.questions)
                }
                AppConstants.NO_MATCHING_RESULT -> onNoMatchingResult()
                AppConstants.FAILED -> onError()
            }
        })
        binding.rvSearchResults.adapter = searchAdapter
        searchAdapter.setOnClickListener(viewHolderClickListener, shareClickListener)
    }

    private val viewHolderClickListener = View.OnClickListener {
        val viewHolder = it.tag as RecyclerView.ViewHolder
        val position = viewHolder.bindingAdapterPosition
        Intent(this, AnswerActivity::class.java).apply {
            val currentQuestion = questions!![position]
            putExtra(AppConstants.EXTRA_QUESTION_TITLE, currentQuestion.title)
            putExtra(AppConstants.EXTRA_QUESTION_DATE, AppUtils.toNormalDate(currentQuestion.creationDate!!.toLong()))
            putExtra(AppConstants.EXTRA_QUESTION_FULL_TEXT, currentQuestion.body)
            putExtra(AppConstants.EXTRA_QUESTION_ANSWERS_COUNT, currentQuestion.answerCount)
            putExtra(AppConstants.EXTRA_QUESTION_ID, currentQuestion.questionId)
            putExtra(AppConstants.EXTRA_QUESTION_VOTES_COUNT, currentQuestion.score)
            putExtra(AppConstants.EXTRA_QUESTION_LINK, currentQuestion.link)
            val questionOwner = currentQuestion.owner
            if (questionOwner != null) {
                putExtra(AppConstants.EXTRA_QUESTION_NAME, questionOwner.displayName)
                putExtra(AppConstants.EXTRA_AVATAR_ADDRESS, questionOwner.profileImage)
                putExtra(AppConstants.EXTRA_QUESTION_OWNER_LINK, questionOwner.link)
            }
            startActivity(this)
        }
    }

    private val shareClickListener = View.OnClickListener { v ->
        val currentQuestion = v.tag as? Question
        if (currentQuestion != null) {
            AppUtils.shareContent(currentQuestion.link!!, this)
        }
    }

    private fun setQuery() {
        searchViewModel.setQuery(searchInput!!)
    }

    private fun onLoading() = binding.apply {
        searchPbFetchData.visibility = View.VISIBLE
        rvSearchResults.visibility = View.INVISIBLE
        searchTvError.visibility = View.INVISIBLE
    }

    private fun onLoaded() = binding.apply {
        searchPbFetchData.visibility = View.INVISIBLE
        rvSearchResults.visibility = View.VISIBLE
        searchTvError.visibility = View.INVISIBLE
    }

    private fun onNoMatchingResult() = binding.apply {
        searchPbFetchData.visibility = View.INVISIBLE
        rvSearchResults.visibility = View.INVISIBLE
        searchTvError.visibility = View.VISIBLE
        searchTvError.setText(R.string.no_matching_result)
    }

    private fun onError() = binding.apply {
        searchPbFetchData.visibility = View.INVISIBLE
        rvSearchResults.visibility = View.INVISIBLE
        searchTvError.visibility = View.VISIBLE
        searchTvError.setText(R.string.network_error_message)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        super.onBackPressed()
        finish()
    }
}